esp-hal/src/dma/mod.rs

```diff
@@ -607,32 +575,69 @@ macro_rules! dma_descriptors_chunk_size {
-        // these will check for size at compile time
-        const _: () = ::core::assert!($chunk_size <= 4095, "chunk size must be <= 4095");
-        const _: () = ::core::assert!($chunk_size > 0, "chunk size must be > 0");
+        $crate::dma_descriptors_impl!($rx_size, $tx_size, $chunk_size, is_circular = true)
+    }};
 
-        const rx_descriptor_len: usize = if $rx_size > $chunk_size * 2 {
-            ($rx_size + $chunk_size - 1) / $chunk_size
-        } else {
-            3
-        };
+    ($size:expr, $chunk_size:expr) => {
+        $crate::dma_circular_descriptors_chunk_size!($size, $size, $chunk_size)
+    };
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! dma_buffers_impl {
+    ($rx_size:expr, $tx_size:expr, $chunk_size:expr, is_circular = $circular:tt) => {{
+        let rx = $crate::dma_buffers_impl!($rx_size, $chunk_size, is_circular = $circular);
+        let tx = $crate::dma_buffers_impl!($tx_size, $chunk_size, is_circular = $circular);
+        (rx.0, rx.1, tx.0, tx.1)
+    }};
 
-        const tx_descriptor_len: usize = if $tx_size > $chunk_size * 2 {
-            ($tx_size + $chunk_size - 1) / $chunk_size
+    ($size:expr, $chunk_size:expr, is_circular = $circular:tt) => {{
+        $crate::declare_aligned_dma_buffer!(BUFFER, $size);
+
+        unsafe {
+            (
+                $crate::as_mut_byte_array!(BUFFER, $size),
+                $crate::dma_descriptors_impl!($size, $chunk_size, is_circular = $circular),
+            )
+        }
+    }};
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! dma_descriptors_impl {
+    (@count $size:expr, $chunk_size:expr, is_circular = true) => {
+        if $size > $chunk_size * 2 {
+            ($size as usize).div_ceil($chunk_size)
         } else {
             3
-        };
+        }
+    };
 
-        static mut RX_DESCRIPTORS: [$crate::dma::DmaDescriptor; rx_descriptor_len] =
-            [$crate::dma::DmaDescriptor::EMPTY; rx_descriptor_len];
-        static mut TX_DESCRIPTORS: [$crate::dma::DmaDescriptor; tx_descriptor_len] =
-            [$crate::dma::DmaDescriptor::EMPTY; tx_descriptor_len];
-        unsafe { (&mut RX_DESCRIPTORS, &mut TX_DESCRIPTORS) }
+    (@count $size:expr, $chunk_size:expr, is_circular = false) => {
+        ($size as usize).div_ceil($chunk_size)
+    };
+
+    ($rx_size:expr, $tx_size:expr, $chunk_size:expr, is_circular = $circular:tt) => {{
+        let rx = $crate::dma_descriptors_impl!($rx_size, $chunk_size, is_circular = $circular);
+        let tx = $crate::dma_descriptors_impl!($tx_size, $chunk_size, is_circular = $circular);
+        (rx, tx)
     }};
 
-    ($size:expr, $chunk_size:expr) => {
-        $crate::dma_circular_descriptors_chunk_size!($size, $size, $chunk_size)
-    };
+    ($size:expr, $chunk_size:expr, is_circular = $circular:tt) => {{
+        // these will check for size at compile time
+        const _: () = ::core::assert!($chunk_size <= 4095, "chunk size must be <= 4095");
```

Comment:
I guess this can be done now
```suggestion
        const { ::core::assert!($chunk_size <= 4095, "chunk size must be <= 4095"); }
```